src/liballoc/collections/btree/navigate.rs

```diff
@@ -192,8 +139,10 @@ impl<K, V> Handle<NodeRef<marker::Owned, K, V, marker::Leaf>, marker::Edge> {
     ///   if the two preconditions above hold.
     /// - Using the updated handle may well invalidate the returned references.
     pub unsafe fn next_unchecked(&mut self) -> (K, V) {
-        let (next_edge, k, v) = next_unchecked_deallocating(ptr::read(self));
-        *self = next_edge;
+        let kv = next_kv_unchecked_dealloc(ptr::read(self));
+        let k = ptr::read(kv.reborrow().into_kv().0);
+        let v = ptr::read(kv.reborrow().into_kv().1);
+        *self = kv.next_leaf_edge();
```

Comment:
Could we have a function like this to replace the ptr::read/*self in all of these functions? It won't help with all the pointer reads but would make this code much clearer I think.

```rust
/// This replaces the value behind the `v` unique reference by calling the
/// relevant function.
///
/// Safety: The change closure must not panic.
unsafe fn replace<R>(v: &mut T, change: impl FnOnce(T) -> (T, R)) -> R {
    let value = ptr::read(v);
    let (new_value, ret) = change(value);
    ptr::write(v, new_value);
    ret
}
```

Reply:
I can try to wiggle such a function in there and keep the compiler and Miri happy, but it's not quite clear to me. I'm still wondering why that `kv.reborrow().into_kv()` is written twice.

Reply:
To be clear, the reborrow().into_kv() would stay exactly the same with the new function, it's just that the ptr::read/*self wrapping it would go into the function vs. being duplicated everywhere.

Reply:
I understand the purpose of the function now.